QMLComponents/controls/radiobuttonsgroupbase.cpp

```diff
@@ -33,50 +32,49 @@ RadioButtonsGroupBase::RadioButtonsGroupBase(QQuickItem* item)
 
 void RadioButtonsGroupBase::setUp()
 {
-	QList<RadioButtonBase* > buttons;
-	_getRadioButtons(this, buttons);
-	QVariant buttonGroup = property("buttonGroup");
-
-	connect(this,	&RadioButtonsGroupBase::clicked, this,	&RadioButtonsGroupBase::clickedSlot);
-
-	for (RadioButtonBase* button: buttons)
-	{
-		const QString& controlName = button->name();
-		if (controlName.isEmpty())
-			addControlError(tr("A RadioButton inside RadioButtonGroup element (name: %1) does not have any name").arg(name()));
-		else
-		{
-			_buttons[controlName] = button;
-			if (button->property("checked").toBool())	setValue(controlName);
-			button->setProperty("buttonGroup", buttonGroup);
-		}
-	}
+	JASPControl::setUp();
+}
 
-	if (value().isEmpty() && _buttons.size() > 0)
+void RadioButtonsGroupBase::registerRadioButton(RadioButtonBase* button)
+{
+	const QString& controlName = button->name();
+	if (controlName.isEmpty())
+		addControlError(tr("A RadioButton inside RadioButtonGroup element (name: %1) does not have any name").arg(name()));
+	else
 	{
-		RadioButtonBase* firstButton = _buttons.values()[0];
-		Log::log() << "No checked button found in radio buttons " << name() << ". First one (" << firstButton->name() << ") is checked per default" << std::endl;
-		_setCheckedButton(firstButton);
+		if (_valueButtonMap.size() == 0)
+			_setCheckedButton(button);
+		if (button->property("checked").toBool())
+			_setCheckedButton(button);
+		_valueButtonMap[controlName] = button;
+		_buttonValueMap[button] = controlName;
+		emit buttonsChanged();
 	}
+}
 
-	JASPControl::setUp();
+void RadioButtonsGroupBase::unregisterRadioButton(RadioButtonBase* button)
+{
+	_valueButtonMap.remove(button->name());
+	_buttonValueMap.remove(button);
+	if (_value == button->name() && _valueButtonMap.size() > 0)
+		_setCheckedButton(_valueButtonMap.first());
+	emit buttonsChanged();
 }
 
-void RadioButtonsGroupBase::_getRadioButtons(QQuickItem* item, QList<RadioButtonBase *> &buttons) {
-	for (QQuickItem* child : item->childItems())
+
+void RadioButtonsGroupBase::radioButtonValueChanged(RadioButtonBase *button)
+{
+	if (_buttonValueMap.contains(button))
 	{
-		JASPControl* jaspControl = qobject_cast<JASPControl*>(child);
-		if (jaspControl)
-		{
-			ControlType controlType = jaspControl->controlType();
-			if (controlType == ControlType::RadioButton)
-				buttons.append(qobject_cast<RadioButtonBase*>(jaspControl));
-			else if (controlType != ControlType::RadioButtonGroup)
-				_getRadioButtons(child, buttons);
-		}
-		else
-			_getRadioButtons(child, buttons);
-	}	
+		QString oldValue = _buttonValueMap[button];
+		_valueButtonMap.remove(oldValue);
```

Comment:
Why store the value in a map?

When you need the value you can just get it from the RadioButton through its pointer, no?

Reply:
The value -> button map is for the bindTo function. I suppose I could have done a linear scan over a set instead but a map is most natural.
The second map is a bit silly but when the name changes I need to have to old name() value to check if it is the current group  value.
There is probably a nicer way but I didnt want to alter the changeName logic in jaspControl so I took the easy route.